installer/block_device.go

```diff
@@ -195,3 +252,155 @@ func (bd *BlockDevice) SectorSize() (int, error) {
+// method, with the next len([]byte) bytes (considered a frame) on the block
+// device, and if they match, the write is discarded. The lingo is that only
+// dirty frames are written. Clean ones are discarded.
+type OptimizedBlockDeviceWriter struct {
+	blockDevice BlockDevicer
+	totalFrames int
+	dirtyFrames int
+}
+
+// Write only write 'dirty' frames.
+// Note: a frame-size is always the size 'len(b)'
+func (bd *OptimizedBlockDeviceWriter) Write(b []byte) (n int, err error) {
+
+	log.Infof("Optimized block device writer: len(b): %d, b: %s", len(b), string(b[:2]))
+	log.Infof("Framesize: %d", len(b))
+
+	frameSize := int64(len(b))
+	payloadBuf := make([]byte, frameSize)
+
+	//
+	// Read len(b) bytes from the block-device
+	//
+	n, err = io.ReadFull(bd.blockDevice, payloadBuf)
+	if err != nil {
+		log.Errorf("Failed to read a full frame of size: %d from the block-device", err)
+		return 0, err
+	}
+
+	//
+	// Write the frame if it is dirty.
+	//
+	if !bytes.Equal(payloadBuf, b) {
+		log.Info("Dirty frame detected. Rewriting...")
+		// In order to write, we need to seek back to
+		// the start of the chunk.
+		if _, err = bd.blockDevice.Seek(-int64(frameSize), io.SeekCurrent); err != nil {
+			log.Errorf("Failed to seek back to the start of the frame")
+			return 0, err
+		}
+		bd.totalFrames += 1
+		bd.dirtyFrames += 1
+		return bd.blockDevice.Write(b)
+	}
+
+	log.Infof("Skipping writing of clean frame. Reporting: %d bytes as written", n)
+	// No need to write a clean frame
+	bd.totalFrames += 1
+	return n, err
+}
+
+func (obw *OptimizedBlockDeviceWriter) Close() error {
+	s := "The optimized block-device writer wrote a total of %d frames, " +
+		"where %d frames did need to be rewritten"
+	log.Infof(s, obw.totalFrames, obw.dirtyFrames)
+	return obw.blockDevice.Close()
+}
+
+// FlushingWriter is a wrapper around an *os.File which forces a Sync() to occur
+// every FlushIntervalBytes.
+type FlushingWriter struct {
+	*os.File
+	FlushIntervalBytes    uint64
+	unflushedBytesWritten uint64
+}
+
+// NewFlushingWriter returns a FlushingWriter which wraps the provided
+// WriteSyncer and automatically flushes (calls Sync()) each time the specified
+// number of bytes is written.
+// Setting flushIntervalBytes == 0 causes Sync() to be called after every Write().
+func NewFlushingWriter(wf *os.File, flushIntervalBytes uint64) *FlushingWriter {
+	return &FlushingWriter{
+		File:                  wf,
+		FlushIntervalBytes:    flushIntervalBytes,
+		unflushedBytesWritten: 0,
+	}
+}
+
+func (fw *FlushingWriter) Write(p []byte) (int, error) {
+	log.Infof("Writing with the flush-writer: %s", string(p[:2]))
+	rv, err := fw.File.Write(p)
+
+	fw.unflushedBytesWritten += uint64(rv)
+
+	if err != nil {
+		return rv, err
+	} else if fw.unflushedBytesWritten >= fw.FlushIntervalBytes {
+		err = fw.Sync()
+	}
+
+	return rv, err
+}
+
+func (fw *FlushingWriter) Sync() error {
+	err := fw.File.Sync()
+	fw.unflushedBytesWritten = 0
+	return err
+}
```

Comment:
Confused. Isn't `FlushingWriter` the `BlockDevicer` at the bottom of the layers? Doesn't it need to support `Seek()` then?

Reply:
It does support seek by composition

Reply:
Ah right, I missed that.